src/ripple/unl/tests/qalloc.h

```diff
@@ -0,0 +1,396 @@
+    using pointer = T*;
+    using const_pointer = T const*;
+    using reference = typename
+        std::add_lvalue_reference<T>::type;
+    using const_reference = typename
+        std::add_lvalue_reference<T const>::type;
+
+    template <class U>
+    struct rebind
+    {
+        using other = qalloc_type<U>;
+    };
+
+    qalloc_type (qalloc_type const&) = default;
+    qalloc_type& operator= (qalloc_type const&) = default;
+#ifndef _MSC_VER
+    qalloc_type (qalloc_type&& other) = default;
+    qalloc_type& operator= (qalloc_type&&) = default;
+#endif
+
+    qalloc_type();
+
+    template <class U>
+    qalloc_type (qalloc_type<U> const& u);
+
+    T*
+    allocate (std::size_t n);
+
+    template <class U>
+    U*
+    alloc (std::size_t n);
+
+    void
+    deallocate (T* p, std::size_t n);
+
+    template <class U>
+    void
+    dealloc (U* p, std::size_t n);
+
+    void
+    destroy (T* t);
+
+    template <class U>
+    bool
+    operator== (qalloc_type<U> const& u);
+
+    template <class U>
+    bool
+    operator!= (qalloc_type<U> const& u);
+};
+
+/** Allocator optimized for delete in temporal order.
+
+    This allocator is optimized for the case where objects
+    are deleted in approximately the same order that they
+    were created.
+*/
+using qalloc = qalloc_type<int>;
+
+//------------------------------------------------------------------------------
+
+namespace detail {
+
+template <class _>
+qalloc_impl<_>::block::block (std::size_t bytes)
+    : bytes_ (bytes - sizeof(*this))
+    , remain_ (bytes_)
+    , free_ (reinterpret_cast<
+        std::uint8_t*>(this + 1))
+{
+}
+
+template <class _>
+void*
+qalloc_impl<_>::block::allocate(
+    std::size_t bytes, std::size_t align)
+{
+//align = 16;
+    align = std::max(align,
+        std::alignment_of<block*>::value);
+    auto pad = [](void const* p, std::size_t align)
+    {
+        auto const i = reinterpret_cast<
+            std::uintptr_t>(p);
+        return (align - (i % align)) % align;
+    };
+
+    auto const n0 =
+        pad(free_ + sizeof(block*), align);
+    auto const n1 =
+        n0 + sizeof(block*) + bytes;
+    if (remain_ < n1)
+        return nullptr;
+    auto p = reinterpret_cast<block**>(
+        free_ + n0 + sizeof(block*));
+    assert(pad(p - 1,
+        std::alignment_of<block*>::value) == 0);
+    p[-1] = this;
+    ++count_;
+    free_ += n1;
+    remain_ -= n1;
+    return p;
+}
+
+template <class _>
+bool
+qalloc_impl<_>::block::deallocate()
+{
+    --count_;
+    if (count_ > 0)
+        return false;
+    remain_ = bytes_;
+    free_ = reinterpret_cast<
+        std::uint8_t*>(this + 1);
+    return true;
+}
+
+template <class _>
+template <class U>
+inline
+auto
+qalloc_impl<_>::block::stash (U const* u) ->
+    block*&
+{
+    static auto const a =
+        std::alignment_of<block*>::value;
+    return *reinterpret_cast<block**>(
+        a * ((reinterpret_cast<std::uintptr_t>(u) - sizeof(block*)) / a));
```

Comment:
given the implementation of alloc, shouldn't this just be: `reinterpret_cast<std::uintptr_t>(u) - sizeof(block*)`